### ISSUE TITLE
nfa vs dfa mismatch

### ISSUE THREAD

**Issue by BurntSushi** [rust-lang/regex#345]
This program does not pass all asserts:

```rust
extern crate regex;

use regex::internal::ExecBuilder;

pub const HAYSTACK_BYTES: &'static [u8] = b"\x3D\x86\x3D\x79";

fn main() {
    let pattern = r"=\b";
    let haystack = &HAYSTACK_BYTES;
    let re0 = ExecBuilder::new(pattern).build().unwrap().into_byte_regex();
    let re1 = ExecBuilder::new(pattern).nfa().build().unwrap().into_byte_regex();
    assert_eq!(re0.is_match(haystack), re1.is_match(haystack));
    assert_eq!(re0.find_iter(haystack).collect::<Vec<_>>(),
               re1.find_iter(haystack).collect::<Vec<_>>());
}
```

Found by @lukaslueg in #321

**Comment by BurntSushi** [comment]
Looks like this is probably a discrepancy in how word boundaries and invalid UTF-8 are handled in the matching engines. Notably, the pattern `=(?-u:\b)` also fails, but the `is_match` results are inverted.

**Comment by SeanRBurton** [comment]
A similar example:

```
fn main() {
    let re0 = ExecBuilder::new("$\\B(?m:^)").only_utf8(false).build().unwrap();
    let re1 = ExecBuilder::new("$\\B(?m:^)").only_utf8(false).nfa().build().unwrap();
    let re0 = re0.into_regex();
    let re1 = re1.into_regex();
    assert!(re0.is_match("\n") == re1.is_match("\n"));
}
```

**Comment by BurntSushi** [comment]
This appears to no longer be an issue. I suspect it was fixed by #561.
